### PR TITLE
Add a visibility limit to the map

### PR DESCRIPTION
**Pull request by rutgerkok**
This change adds a visibility limit to the map, so that only a limited region is rendered:

Pl3xmap in a limited region

Right now, it's kind of a proof-of-concept. The implementation works and I think it can be merged in this state. However, it currently only supports defining one circle as the visibility limit. The limit is disabled by default. The following configuration options are available:

**config.yml**
```yaml
world-settings:
  default:
    map:
      ...
      visibility-limit:
        enabled: true
        radius: 200
        center-x: 3782
        center-z: 6844
```

The implementation is centered around the `VisibilityLimit` class. That class decides what regions, chunks and block columns are visible on the map. If an entire region or chunk isn't visible, then it is skipped altogether. The class has been adapted from [Topographica](https://github.com/rutgerkok/Topographica), another 2D livemap plugin, written by me. I intend to discontinue that plugin because Pl3xmap produces more beautiful images, because the web interface has more functionality and because Pl3xmap has faster rendering as it uses Paper's chunk loading API.

In the future, I can imagine that the `VisibilityLimit` class will loop through a list of shapes, and query each of them whether that area of the map is visible. Right now, the shape of a circle is directly coded into the class. For reference, Dynmap supports both circular and rectangular visibility limits, and also hidden limits, which hide everything within the limits.

Two other commits were included to make me able to work on this repo. https://github.com/pl3xgaming/Pl3xMap/commit/ca67ebbbd1e68c5d4c35117ba0cfc06aa8478d61 adds some Eclipse files to the .gitignore (just like the IntelliJ files are ignored), while https://github.com/pl3xgaming/Pl3xMap/commit/6f1812586774bd49e21fc9bccca1179282f15cf5 fixes a case where the code [didn't follow the Java Language Specification](https://stackoverflow.com/questions/29097160/why-are-nonnull-annotation-on-fqn-not-allowed), causing the Eclipse Compiler to emit a compile error.

### CODE REVIEW

**Comment by pop4959** [comment]
I personally love this idea. I do think it should be adapted to support a square render shape at the very least, seeing that the vanilla Minecraft world border is square shaped and [the most common shapes in border plugins](https://bstats.org/plugin/bukkit/ChunkyBorder/8953#borderShape) (at least mine) are square then circle. This could relatively easily be implemented by just taking the radius in the config to be the apothem. Although it would probably be best to refactor the code a bit more too to make it possible to implement other visibility regions later (and maybe with API).

**Comment by rutgerkok** [comment]
That was the kind of feedback I was looking for, @pop4959 !

If Billy approves this idea at all, then I guess I can extend the config file so that it looks like this:


```yml
world-settings:
  default:
    map:
      ...
      visibility-limits:
      - shape: circle
        radius: 200
        center-x: 3782
        center-z: 6844
      - shape: rectangle
        min-x: 4000
        max-x: 5000
        min-z: 1000
        max-z: 2000
```

This would then show everything that is either within the circle or within the rectangle.

An API to register more shapes would be nice. Maybe  `minecraft-worldborder`  or `chunkyborder` (via an API) could even be added as a "shape" that automatically follows the bortder, so that the user doesn't need to keep both in sync.

But let's first see what Billy thinks.

**Comment by rutgerkok** [comment]
In an effort to make this more attractive to merge, I've now refactored the code a bit. 😃 

Multiple limits are now possible. By default, the config looks like this:

```yaml
world-settings:
  default:
    map:
      ...
      visibility-limits:
      - shape: world-border
        enabled: false
```

This should make it easy for the server admin to trim the map to the vanilla world border. The fact that it's a list should also provide a hint to the server admin that more limits can be defined. For example:

```yaml
      visibility-limits:
      - type: world-border
        enabled: true
      - type: circle
        radius: 100
        center-x: 3782
        center-z: 6844
      - type: rectangle
        min-x: 3770
        max-x: 3790
        min-z: 6650
        max-z: 6850
```

results in:

Multiple limits

Third party plugins (CC @pop4959 ) can add their own limits through the API:

`mapWorld.visibilityLimit().getShapes().add(...)`

You need to add an instance of `VisibilityShape` to this list. `VisibilityShape` is an interface that allows Pl3xMap to query the shape for what columns/chunks/regions should be drawn.

**Comment by pop4959** [comment]
Neat, thanks so much for the API consideration! I probably would have been willing to add this myself if you hadn't had the time for it. Hopefully Billy finds time to review this sometime soon, because this would be a really cool addition.

File: plugin/src/main/java/net/pl3x/map/plugin/visibilitylimit/VisibilityLimit.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        return this.shapes;
+    }
+
+    @Override
+    public boolean isWithinLimit(int blockX, int blockZ) {
+        return this.shouldRenderColumn(blockX, blockZ);
+    }
+
+    public void parse(List<Map<String, Object>> configLimits) {
+        this.shapes .clear();
```
Comment (BillyGalbreath) [inline plugin/src/main/java/net/pl3x/map/plugin/visibilitylimit/VisibilityLimit.java:75]:
```suggestion
        this.shapes.clear();
```

**Review by BillyGalbreath — APPROVED** [review]
Looks good to me. Just the 1 nitpick.

**Comment by rutgerkok** [comment]
Thanks for reviewing! I fixed the extra space.

**Pull request merged**